source/common/network/connection_impl.cc

```diff
@@ -323,8 +326,9 @@ void ConnectionImpl::readDisable(bool disable) {
     file_event_->setEnabled(Event::FileReadyType::Read | Event::FileReadyType::Write);
     // If the connection has data buffered there's no guarantee there's also data in the kernel
     // which will kick off the filter chain. Instead fake an event to make sure the buffered data
-    // gets processed regardless.
+    // gets processed regardless and ensure that we push it up via onRead.
     if (read_buffer_.length() > 0) {
+      force_on_read_ = true;
```

Comment:
So we call this on the connection in two places, here and setReadBufferReady()

Looking at call sites of setReadBufferReady it looks like we're doing that where we actually want to resume reading from the socket (and don't want to cause a spurious onRead if there's no further data).  I think I'm inclined to just leave this as-is, but if we think future users of the network connection might not fully read data and use setReadBufferReady to trigger the dispatch (rather than readEnable/Disable as we do here), we may want to add a boolean to force on read there as well.